lib/utils/selector.js

```diff
@@ -0,0 +1,582 @@
+  }
+}
+/**
+ * Convert Selector child nodes to VElementMatcher
+ * @param {ChildNode[]} selectorChildren
+ * @returns {VElementMatcher}
+ */
+function selectorToVElementMatcher(selectorChildren) {
+  const nodes = [...selectorChildren]
+  let node = nodes.shift()
+  /**
+   * @type {VElementMatcher | null}
+   */
+  let result = null
+  while (node) {
+    if (node.type === 'combinator') {
+      const combinator = node.value
+      node = nodes.shift()
+      if (!node) {
+        throw new SelectorError(`Expected selector after '${combinator}'.`)
+      }
+      if (node.type === 'combinator') {
+        throw new SelectorError(`Unexpected combinator '${node.value}'.`)
+      }
+      const right = nodeToVElementMatcher(node)
+      result = combination(
+        result ||
+          // for :has()
+          ((element, subject) => element === subject),
+        combinator,
+        right
+      )
+    } else {
+      const sel = nodeToVElementMatcher(node)
+      result = result ? compound(result, sel) : sel
+    }
+    node = nodes.shift()
+  }
+  if (!result) {
+    throw new SelectorError(`Unexpected empty selector.`)
+  }
+  return result
+}
+
+/**
+ * @param {VElementMatcher} left
+ * @param {string} combinator
+ * @param {VElementMatcher} right
+ * @returns {VElementMatcher}
+ */
+function combination(left, combinator, right) {
+  if (!combinator.trim()) {
+    // descendant
+    return (element, subject) => {
+      if (right(element, null)) {
+        let parent = element.parent
+        while (parent.type === 'VElement') {
+          if (left(parent, subject)) {
+            return true
+          }
+          parent = parent.parent
+        }
+      }
+      return false
+    }
+  }
+
+  if (combinator === '>') {
+    // child
+    return (element, subject) => {
+      if (right(element, null)) {
+        const parent = element.parent
+        if (parent.type === 'VElement') {
+          return left(parent, subject)
+        }
+      }
+      return false
+    }
+  }
+
+  if (combinator === '+') {
+    // adjacent
+    return (element, subject) => {
+      if (right(element, null)) {
+        const before = getBeforeElement(element)
+        if (before) {
+          return left(before, subject)
+        }
+      }
+      return false
+    }
+  }
+
+  if (combinator === '~') {
+    // sibling
+    return (element, subject) => {
+      if (right(element, null)) {
+        for (const before of getBeforeElements(element)) {
+          if (left(before, subject)) {
+            return true
+          }
+        }
+      }
+      return false
+    }
+  }
+  throw new SelectorError(`Unknown combinator: ${combinator}.`)
+}
+
+/**
+ * Convert node to VElementMatcher
+ * @param {Exclude<parser.Node, {type:'combinator'|'comment'|'root'|'selector'}>} selector
+ * @returns {VElementMatcher}
+ */
+function nodeToVElementMatcher(selector) {
+  if (selector.type === 'attribute') {
```

Comment:
There are a lot of similar conditions in row here: 

```js
if (selector.type === '…') {
  // …
  return …
}
// similar cases following
```

This might be simplified/clarified to:

```js
switch (selector.type) {
  case '…': {
    // …
    return …
  }
  // similar cases following
}
```